File: detection_ws/src/value_map/value_map/test/test_service.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from seem_ros_interfaces.srv import Panoptic
from cv_bridge import CvBridge
import cv2


class PanopticClientNode(Node):
    def __init__(self):
        super().__init__('panoptic_client_node')
        self.get_logger().info('PanopticClientNode started.')

        # CV Bridge für Image-Konvertierung
        self.bridge = CvBridge()
        self.rgb_image = None

        # Subscriber für RGB
        self.subscription = self.create_subscription(
            Image,
            '/rgb',
            self.image_callback,
            10
        )

        # Service Client
        self.cli = self.create_client(Panoptic, 'panoptic_segmentation')

        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn('Waiting for panoptic_segmentation service...')

        # Timer mit 1 Hz
        self.timer = self.create_timer(0.1, self.timer_callback)

    def image_callback(self, msg):
        self.rgb_image = msg

    def timer_callback(self):
        if self.rgb_image is None:
            self.get_logger().warn('No image received yet.')
            return

        req = Panoptic.Request()
        req.image = self.rgb_image

        self.future = self.cli.call_async(req)
        self.future.add_done_callback(self.handle_response)

    def handle_response(self, future):
        try:
            response = future.result()
            seg_image = self.bridge.imgmsg_to_cv2(response.panoptic_segmentation, desired_encoding='rgb8')
            cv2.imshow("Panoptic Segmentation", seg_image)
            cv2.waitKey(1)
        except Exception as e:
            self.get_logger().error(f'Service call failed: {e}')


def main(args=None):
    rclpy.init(args=args)
    node = PanopticClientNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: explorer_ws/src/value_map/src/test/test_service.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "seem_ros_interfaces/srv/panoptic.hpp"
#include "seem_ros_interfaces/srv/object_segmentation.hpp"
#include "seem_ros_interfaces/srv/semantic_similarity.hpp"
#include "cv_bridge/cv_bridge.h"
#include <opencv2/opencv.hpp>

class ServiceHandler
{
public:
    ServiceHandler(rclcpp::Node::SharedPtr node)
        : node_(node), bridge_(std::make_shared<cv_bridge::CvImage>())
    {
        vlm_namespace_ = node_->declare_parameter<std::string>("vlm_target_namespace", "/seem_ros");
        if (!vlm_namespace_.empty() && vlm_namespace_.back() == '/')
            vlm_namespace_.pop_back();

        auto resolve = [this](const std::string &service) {
            return vlm_namespace_ + "/" + service;
        };

        cli_panoptic_ = node_->create_client<seem_ros_interfaces::srv::Panoptic>(resolve("panoptic_segmentation"));
        cli_object_ = node_->create_client<seem_ros_interfaces::srv::ObjectSegmentation>(resolve("object_segmentation"));
        cli_similarity_ = node_->create_client<seem_ros_interfaces::srv::SemanticSimilarity>(resolve("semantic_similarity"));

        subscription_ = node_->create_subscription<sensor_msgs::msg::Image>(
            "/rgb", 10, std::bind(&ServiceHandler::image_callback, this, std::placeholders::_1));

        RCLCPP_INFO(node_->get_logger(), "ServiceHandler initialized with VLM namespace: %s", vlm_namespace_.c_str());
    }

private:
    void image_callback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        RCLCPP_INFO(node_->get_logger(), "Received image, calling all services...");

        call_panoptic(*msg);
        call_object_segmentation(*msg, "bottle");
        call_semantic_similarity(*msg, "bottle");
    }

    void call_panoptic(const sensor_msgs::msg::Image &image_msg)
    {
        auto request = std::make_shared<seem_ros_interfaces::srv::Panoptic::Request>();
        request->image = image_msg;

        cli_panoptic_->async_send_request(request,
            [this](rclcpp::Client<seem_ros_interfaces::srv::Panoptic>::SharedFuture future) {
                try {
                    auto response = future.get();
                    cv::Mat img = cv_bridge::toCvCopy(response->panoptic_segmentation, "rgb8")->image;
                    cv::imshow("Panoptic Segmentation", img);
                    cv::waitKey(1);
                } catch (const std::exception &e) {
                    RCLCPP_ERROR(node_->get_logger(), "Panoptic response error: %s", e.what());
                }
            });
    }

    void call_object_segmentation(const sensor_msgs::msg::Image &image_msg, const std::string &query)
    {
        auto request = std::make_shared<seem_ros_interfaces::srv::ObjectSegmentation::Request>();
        request->image = image_msg;
        request->query = query;

        cli_object_->async_send_request(request,
            [this](rclcpp::Client<seem_ros_interfaces::srv::ObjectSegmentation>::SharedFuture future) {
                try {
                    auto response = future.get();
                    cv::Mat img = cv_bridge::toCvCopy(response->segmented_image, "rgb8")->image;
                    cv::imshow("Object Segmentation", img);
                    cv::waitKey(1);
                } catch (const std::exception &e) {
                    RCLCPP_ERROR(node_->get_logger(), "Object response error: %s", e.what());
                }
            });
    }

    void call_semantic_similarity(const sensor_msgs::msg::Image &image_msg, const std::string &query)
    {
        auto request = std::make_shared<seem_ros_interfaces::srv::SemanticSimilarity::Request>();
        request->image = image_msg;
        request->query = query;

        cli_similarity_->async_send_request(request,
            [this](rclcpp::Client<seem_ros_interfaces::srv::SemanticSimilarity>::SharedFuture future) {
                try {
                    auto response = future.get();
                    RCLCPP_INFO(node_->get_logger(), "Semantic similarity score: %.3f", response->score);
                } catch (const std::exception &e) {
                    RCLCPP_ERROR(node_->get_logger(), "Similarity response error: %s", e.what());
                }
            });
    }

    rclcpp::Node::SharedPtr node_;
    std::string vlm_namespace_;
    std::shared_ptr<cv_bridge::CvImage> bridge_;

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;

    rclcpp::Client<seem_ros_interfaces::srv::Panoptic>::SharedPtr cli_panoptic_;
    rclcpp::Client<seem_ros_interfaces::srv::ObjectSegmentation>::SharedPtr cli_object_;
    rclcpp::Client<seem_ros_interfaces::srv::SemanticSimilarity>::SharedPtr cli_similarity_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<rclcpp::Node>("test_service_node");
    auto handler = std::make_shared<ServiceHandler>(node);

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
